moveit_core/robot_model/src/planar_joint_model.cpp

```diff
@@ -139,39 +141,128 @@ void PlanarJointModel::getVariableRandomPositionsNearBy(random_numbers::RandomNu
+  const double angle_straight_diff = std::hypot(dx, dy) > min_translational_distance ?
+                                         angles::shortest_angular_distance(from[2], std::atan2(dy, dx)) :
+                                         0.0;
+  const double angle_backward_diff =
+      angles::normalize_angle(angle_straight_diff - boost::math::constants::pi<double>());
+  const double move_straight_cost =
+      std::abs(angle_straight_diff) + std::abs(angles::shortest_angular_distance(from[2] + angle_straight_diff, to[2]));
+  const double move_backward_cost =
+      std::abs(angle_backward_diff) + std::abs(angles::shortest_angular_distance(from[2] + angle_backward_diff, to[2]));
+  if (move_straight_cost <= move_backward_cost)
+  {
+    initial_turn = angle_straight_diff;
+  }
+  else
+  {
+    initial_turn = angle_backward_diff;
+  }
+  drive_angle = from[2] + initial_turn;
+  final_turn = angles::shortest_angular_distance(drive_angle, to[2]);
+}
+
 void PlanarJointModel::interpolate(const double* from, const double* to, const double t, double* state) const
 {
-  // interpolate position
-  state[0] = from[0] + (to[0] - from[0]) * t;
-  state[1] = from[1] + (to[1] - from[1]) * t;
+  if (motion_model_ == HOLONOMIC)
+  {
+    // interpolate position
+    state[0] = from[0] + (to[0] - from[0]) * t;
+    state[1] = from[1] + (to[1] - from[1]) * t;
 
-  // interpolate angle
-  double diff = to[2] - from[2];
-  if (fabs(diff) <= boost::math::constants::pi<double>())
-    state[2] = from[2] + diff * t;
-  else
+    // interpolate angle
+    double diff = to[2] - from[2];
+    if (fabs(diff) <= boost::math::constants::pi<double>())
+      state[2] = from[2] + diff * t;
+    else
+    {
+      if (diff > 0.0)
+        diff = 2.0 * boost::math::constants::pi<double>() - diff;
+      else
+        diff = -2.0 * boost::math::constants::pi<double>() - diff;
+      state[2] = from[2] - diff * t;
+      // input states are within bounds, so the following check is sufficient
+      if (state[2] > boost::math::constants::pi<double>())
+        state[2] -= 2.0 * boost::math::constants::pi<double>();
+      else if (state[2] < -boost::math::constants::pi<double>())
+        state[2] += 2.0 * boost::math::constants::pi<double>();
+    }
+  }
+  else if (motion_model_ == DIFF_DRIVE)
   {
-    if (diff > 0.0)
-      diff = 2.0 * boost::math::constants::pi<double>() - diff;
+    double dx, dy, initial_turn, drive_angle, final_turn;
+    computeTurnDriveTurnGeometry(from, to, min_translational_distance_, dx, dy, initial_turn, drive_angle, final_turn);
+
+    double initial_d = fabs(initial_turn) * angular_distance_weight_;
+    double drive_d = hypot(dx, dy);
+    double final_d = fabs(final_turn) * angular_distance_weight_;
+
+    double total_d = initial_d + drive_d + final_d;
+
+    double initial_frac = initial_d / total_d;
+    double drive_frac = drive_d / total_d;
+    double final_frac = final_d / total_d;
+
+    double pct;
```

Comment:
would prefer "percent" so I don't have to spend 2 seconds guessing what the variable means